jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java

```diff
@@ -356,28 +362,77 @@ public ContextHandler createContextHandler(final App app) throws Exception
-                {
-                    Properties p = new Properties();
-                    p.load(stream);
-                    p.stringPropertyNames().forEach(k -> properties.put(k, p.getProperty(k)));
-                }
+                envPropertyFiles = paths.filter(Files::isRegularFile)
+                    .map(p -> parent.relativize(p))
+                    .filter(p ->
+                    {
+                        String name = p.getName(0).toString();
+                        if (!name.endsWith(".properties"))
+                            return false;
+                        if (!name.startsWith(env))
+                            return false;
+                        return true;
+                    }).sorted().collect(Collectors.toList());
+            }
 
-                String str = properties.get(Deployable.ENVIRONMENT_XML);
-                if (!StringUtil.isEmpty(str))
-                {
-                    Path envXmlPath = Paths.get(str);
-                    if (!envXmlPath.isAbsolute())
-                        envXmlPath = getMonitoredDirResource().getPath().getParent().resolve(envXmlPath);
+            if (LOG.isDebugEnabled())
+                LOG.debug("Environment property files {}", envPropertyFiles);
 
-                    context = applyXml(context, envXmlPath, env, properties);
+            Map<String, Path> envXmlFilenameMap = new HashMap<>();
+            for (Path file : envPropertyFiles)
+            {
+                Path resolvedFile = parent.resolve(file);
+                if (Files.exists(resolvedFile))
+                {
+                    Properties tmp = new Properties();
+                    try (InputStream stream = Files.newInputStream(resolvedFile))
+                    {
+                        tmp.load(stream);
+                        //put each property into our substitution pool
+                        tmp.stringPropertyNames().forEach(k -> properties.put(k, tmp.getProperty(k)));
+                        //extract any properties that name environment context xml files
+                        for (Map.Entry<Object, Object> entry : tmp.entrySet())
+                        {
+                            String name = Objects.toString(entry.getKey(), "");
+                            if (name.startsWith(Deployable.ENVIRONMENT_XML))
+                            {
+                                //ensure all environment context xml files are absolute paths
+                                Path envXmlPath = Paths.get(entry.getValue().toString());
+                                if (!envXmlPath.isAbsolute())
+                                    envXmlPath = getMonitoredDirResource().getPath().getParent().resolve(envXmlPath);
+                                //accumulate all properties that name environment xml files so they can be ordered
+                                envXmlFilenameMap.put(name, envXmlPath);
+                            }
+                        }
```

Comment:
This loop should be done on properties AFTER it has been merged into by all the individual files. You are doing this on each iteration on the tmp file, so you might actually act on a property that is subsequently replaced in a later property file.